### PR TITLE
Provide TableBase::replicate_across_communicator().

### LINKED ISSUE #12008: Allow sharing of data across MPI processes
@tjhei and I were talking about sharing large data chunks across MPI processes that live on the same machine and, consequently, can use shmem. We were specifically interested in making this possible with `InterpolatedTensorProductFunction` and friends. For ASPECT, we will probably have some quite large data tables for initial conditions that require this kind of thing -- think 100s of MB or even a few GB; small enough that they can fit into memory, but large enough that it will be quite inefficient if the data is replicated 64 times on a single machine.

Here's how I think we should address this:
* Underlying all relevant classes is `AlignedVector`. If we could get that class to share its data with others, that would already be a big step forward. My plan is to give this class a function `share_data_across_communicator(MPI_Comm)` in which one creates an `AlignedVector` object on one processor and then calls this function collectively on a sub-communicator of processes that can participate via shared memory. The data is then moved into a shmem arena that can be accessed by all other processes.
* The next level up is `TableBase` that under the hood uses `AlignedVector`. It will get the same function and just pass it on to its underlying aligned vector plus exchange minimal amount of information for the size of the table.
* `InterpolatedTensorProductGridData` currently copies its `Table` of data from the calling site, but it would actually make sense to give this class a separate constructor that *moves* the data. In most cases, one is happy to get rid of the table with which this class was constructed, because it is used only in this one context. Once the class is able to move the data, it should also be able to deal with data that lives in a shmem arena.

### PR DESCRIPTION
**Pull request by bangerth**
Closes #12008 at long last.

/rebuild

### CODE REVIEW

File: include/deal.II/base/function_lib.h [hunk 1]
```diff
@@ -1398,6 +1398,55 @@ namespace Functions
...
+   *
+   * The Table class can help with amortizing this cost by using
+   * shared memory to store the data only as often as necessary -- see the
+   * documentation of the TableBase class. Once one has obtained such a
+   * Table object that uses shared memory to store the data only as often
+   * as is necessary, one has to avoid that the current class *copies*
+   * the table into its own member variable. Rather, it is necessary to
+   * use the *move* constructor of this class to take over ownership of
+   * the table and its shared memory space. This can be achieved using
+   * the following extension of the following code snippet shown in the
```
Comment (masterleinad) [inline include/deal.II/base/function_lib.h:1417]:
```suggestion
   * the following extension of the code snippet shown in the
```

File: include/deal.II/base/function_lib.h [hunk 2]
```diff
@@ -1530,6 +1579,14 @@ namespace Functions
    *
    * @note The use of this class is discussed in step-53.
    *
+   *
+   * <h3>Dealing with large data sets</h3>
+   *
+   * This class supportes the same facilities for dealing with large data sets
```
Comment (masterleinad) [inline include/deal.II/base/function_lib.h:1585]:
```suggestion
   * This class supports the same facilities for dealing with large data sets
```

**Comment by bangerth** [comment]
@masterleinad I've incorporated your suggestions.

**Review by masterleinad — APPROVED** [review]

**Comment by bangerth** [comment]
Ping?

**Pull request merged**